pkg/policy/rule.go

```diff
@@ -54,60 +51,18 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
 	//
-	// Existing filter selects all endpoints, so don't add new endpoints.
-	if existingFilter.AllowsAllAtL3() && len(filterToMerge.Endpoints) > 0 {
-		log.WithFields(logrus.Fields{
-			logfields.EndpointSelector: filterToMerge.Endpoints,
-			"policy":                   existingFilter,
-		}).Debug("skipping L4 filter as the endpoints are already covered")
-
-		// Existing L4Filter already selects all endpoints; if there are no L7 rules
-		// to add from the new rule, then we can just exit because the rule is as
-		// permissive as possible at L3, the L4 information is already contained
-		// within the filter, and there is no L7 metadata to add to the filter.
-		if r.NumRules() == 0 {
-			existingFilter.DerivedFromRules = append(existingFilter.DerivedFromRules, ruleLabels)
-			resMap[key] = existingFilter
-			return 1, nil
-		}
+	// Case 1: Both filters select all endpoints
+	if existingFilter.AllowsAllAtL3() && filterToMerge.AllowsAllAtL3() ||
+		existingFilter.AllowsAllAtL3() && !filterToMerge.AllowsAllAtL3() ||
+		!existingFilter.AllowsAllAtL3() && filterToMerge.AllowsAllAtL3() {
+		existingFilter.Endpoints = api.EndpointSelectorSlice{api.WildcardEndpointSelector}
 	} else {
-		// If new rule allows all endpoints, then allow all endpoints.
-		if len(existingFilter.Endpoints) > 0 && filterToMerge.AllowsAllAtL3() {
-			log.WithFields(logrus.Fields{
-				logfields.EndpointSelector: filterToMerge.Endpoints,
-				"policy":                   existingFilter,
-			}).Debug("new L4 filter applies to all endpoints, making the policy more permissive")
-
-			existingFilter.Endpoints = api.EndpointSelectorSlice{api.WildcardEndpointSelector}
-
-			// If new rule allows all endpoints and does have L7 rules, update
-			// filter's L7Parser.
-			if filterToMerge.L7Parser != ParserTypeNone && existingFilter.L7Parser == ParserTypeNone {
-				existingFilter.L7Parser = filterToMerge.L7Parser
-			}
-		} else {
-			existingFilter.Endpoints = append(existingFilter.Endpoints, endpoints...)
-		}
-	}
-
-	// Now, determine whether we allow all on L4.
-	// Rule we are merging with existing L4Filter has no L7 rules, but applies
-	// to the same L4 port-protocol tuple, which *does* have L7 rules. If a rule
-	// applying to the same port/protocol has no L7 rules, then that equates to
-	// allowing all on L7, so just remove existing L7-related metadata for this
-	// port-proto tuple. Or, if we already have a filter which allows all L4, yet
-	// are trying to add a rule which restricts on L7, do not restrict on L7.
-	if r.NumRules() == 0 || existingFilter.L7Parser == ParserTypeNone {
-		for k := range existingFilter.L7RulesPerEp {
-			delete(existingFilter.L7RulesPerEp, k)
-		}
-		existingFilter.L7Parser = ParserTypeNone
-		existingFilter.DerivedFromRules = append(existingFilter.DerivedFromRules, ruleLabels)
-		resMap[key] = existingFilter
-		return 1, nil
+		// Case 4: no wildcard endpoint selectors in existing filter or in filter
```

Comment:
There are only 2 cases:
Case 4 -> Case 2

Same in `mergeL4EgressPort`.

Reply:
@rlenglet I'll fix these in a follow-up PR.